KVStore & derived classes: design docs & implementation

davidsaada opened the pull request:
This PR includes all design docs for KVStore & derived classes.
It is intended for internal review, before publishing globally.

@dannybenor @yossi2le  @offirko @theamirocohen @geky @flit @sg- @amiraloosh @SenRamakri 
Please review using these quick links: [KVStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/KVStore/KVStore_design.md) [TDBStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/TDBStore/TDBStore_design.md)  [FileSystemStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md) [SecureStore](https://github.com/davidsaada/mbed-os/blob/david_kvstore/docs/design-documents/features/storage/SecureStore/SecureStore_design.md)

Commit:
KVStore & derived classes: design docs & implementation

geky commented on features/storage/kvstore/KVStore.h:
```diff
@@ -0,0 +1,199 @@
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size, size_t offset = 0);
```
We should default `actual_size` to NULL (assuming NULL is supported)

davidsaada replied:
Good idea. Will modify.

geky commented on features/storage/kvstore/KVStore.h:
```diff
@@ -0,0 +1,199 @@
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int iterator_next(iterator_t it, char *key, size_t key_size, size_t *actual_key_size);
```
We should default `actual_key_size` to NULL (assuming NULL is supported)

davidsaada replied:
Actually, I removed this parameter altogether (also from the confluence page). Key is a null terminated string, so actual key size is not required.
This got dismissed right at the time you added your comment. :)

geky replied:
Haha, fair enough

What happens if the name is larger than the buffer?

davidsaada replied:
Should return an error (as truncating the string will return a wrong key). We have a define for max key size, so user should be aware of it when allocating the buffer for it.

geky replied:
Good point, sounds like a plan to me

geky commented on features/storage/kvstore/KVStore.h:
```diff
@@ -0,0 +1,199 @@
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);
```
Should we mention somewhere in the doxygen what characters `key` is limited to?

davidsaada replied:
I'll add it.

davidsaada replied:
Well, going over it, it seems a bit awkward as most APIs have a `key` parameter. Adding this description to each one of them seems a bit too much.

geky replied:
Could it just be "Key. Null terminated string. Characters / and * are reserved."?

geky replied:
Note: This is just a suggestion, not a blocker

dreemkiller replied:
What does set do when bits in `create_flags` are not supported?

geky commented on features/storage/kvstore/KVStore.h:
```diff
@@ -0,0 +1,199 @@
+    NVSTORE_AUTH_ERROR             = -9,
+    NVSTORE_RBP_AUTH_ERROR         = -10,
+    KVSTORE_MAX_KEYS_REACHED       = -11,
+} kvstore_status_e;
```
Just a heads up, but I think these may need to be integrated into the global list of mbed-os error codes before this goes in. How to integrate is probably a question for Senthil.

davidsaada replied:
Right. Wasn't aware of the fact this was added, so will take a look.
I'm very much for it, though I foresee some trouble there. Some return values are surely missing (RBP authentication and write once for instance), while others may be ambiguous. For instance, I guess we will have the same error for corrupt data and for authentication error, though they are not the same. The code also makes a lot of usage of these error codes, so I really hope it won't cause any problems.

geky replied:
I'm also not that familiar with the mbed-os error codes, but I believe some of the idea behind it is that we can add error codes as needed.

geky reviewed (commented):
Working on more review feedback but here are some small notes.

geky commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,378 @@
+### Requirements and assumptions
+
+FileSystemStore assumes the underlying file system qualities for resilience and file validation. This means that if the underlying file system has no protection against power failures, then neither would FileSystemStore have.  
+When initializing this class, it is assumed that the underlying FileSystem is initialized and mounted. 
```
I would expect this to behave the same as a FileSystem using a BlockDevice. In which case we call BlockDevice::init during FileSystem::mount no?

Oh is this because of issues with passing the BlockDevice to mount? I think long term we need to deprecate that function and require BlockDevice is provided in the constructor. Would that enable FileSystemStore::init() calling FileSystem::mount?

This can be pulled off in a backwards compatible manner by automatically mounting similarly to how NVStore inits.

If we do this move, I think we should also standardize the FileSystem/KVStore init functions. Perhaps by adopting init/reset in FileSystem?
- FileSystem::mount => FileSystem::init
- FileSystem::unmount => FileSystem::deinit
- FileSystem::reformat => FileSystem::reset

davidsaada replied:
OK, I think this is a key point to explain, which is also related to your comments below: One of the reasons for using this class is if the user wants to use POSIX APIs *in addition* to the get/set ones. This means that the file system is not dedicated to FileSystemStore, but rather hosts it. Using a dedicated file system for FileSystemStore would require a separate block device (or a slice), which adds overhead and wastes space assuming you want a "regular" file system as well.
So - this is why we need a dedicated directory to host the FileSystemStore files; this is why we assume the file system is already initialized; this is why we can't map the `reset` API to the file system `format` one and so on. If it wasn't clear from the design doc, I'll try to explain it more clearly.

geky replied:
The SlicingBlockDevice is a good comparison. It provides organization, but comes with a code cost. So we provide it as an optional layer.

The current FileSystemStore also provides organization, which is a great feature, but it's not the primary goal of KVStore -> FileSystem. We need a way to mount a FileSystem as a KVStore without additional cost.

I think the best thing to do is introduce KVStore into the FileSystem inheritance chain (see the picture above), and keep the FileSystemStore as is (analogous to the SlicingBlockDevice).

---

> One of the reasons for using this class is if the user wants to use POSIX APIs in addition to the get/set ones. This means that the file system is not dedicated to FileSystemStore, but rather hosts it.

Sorry if I'm missing something, but I'm not sure how these statements relate. One example that has come up is MCC + firmware images (this was rejected for other reasons, but is a valid use case). In this case, a dedicated FileSystem/KVStore makes perfect sense. The KVStore API provides the storage for MCC, and the user can create the firmware images through the FileSystem APIs with a name that does not conflict with the KV pairs. Adding the layer for hosting complicated the user interface.

> Using a dedicated file system for FileSystemStore would require a separate block device (or a slice), which adds overhead and wastes space assuming you want a "regular" file system as well.

That's true. On the other hand, maybe a user wants their KV store isolated from the file system to avoid wear-amplification? Maybe they need their KV store to last a long time independently of the file system. Note it would still be advantageous to use a filesystem in both slices because of code sharing.

We don't need to out-think our users. We should be providing flexible options in the code and educating them on best practices in the docs.

davidsaada replied:
@geky I'm probably missing big time here. Maybe your drawing confuses me a bit, especially the line between `FileSystem` and `KVStore`. Do you suggest that `FileSystem` base class should be modified to start inheriting from `KVStore` now? 
In addition, I'm not sure what state we're holding in `FileSystemStore`. The containing directory name is a *fixed* one, with a unique name (say "$FileSystemStore$"). It doesn't need any dynamic allocation whatsoever, nor hold a any state. I think I need more clarifications to this concept before I can go on and refer to the rest of your comments.

geky replied:
Ah,

> Do you suggest that FileSystem base class should be modified to start inheriting from KVStore now?

Yep. This would tie together the two high-level storage classes. The layer from KVStore::get -> FileSystem::file_read can be implemented as default functions in the FileSystem base class. C++ compilers do a good job of optimizing these out if they are unused.

---

> The containing directory name is a _fixed_ one, with a unique name (say "$FileSystemStore$").

Ah thanks, I missed that. So it's impossible to have two independent FileSystemStore instances on a single file system?

> It doesn't need any dynamic allocation whatsoever, nor hold a any state.

How does FileSystemStore construct the new path for the key? Don't we need to convert from "key" to "$FileSystemStore$/key"?

---

The comparison of FileSystemStore and SlicingBlockDevice was with my incorrect assumption that FileSystemStore takes a user-provided directory name. In that case FileSystemStore provides an additional organization feature.

If not, I don't think FileSystemStore is necessary assuming this logic moves to the defaults in FileSystem.

davidsaada replied:
OK, I think I understand now.
I believe this approach has a few problems. You basically want to add all KVStore APIs to FileSystem class, having all file systems inheriting this functionality. In order for this to happen, you will need to implement all these APIs in the FileSysem class. However, FileSystem is just an interface class, meaning that ALL APIs will need implementation in this header file. These are no one liners. The iterator next API, for instance, requires a loop in order to see if the file name matches the prefix. Incremental set APIs have some logic to handle the handles (no pun intended) and so do the iterator ones. Offset argument in the get function also requires some logic, even if a simple one. This approach also makes me understand your other comments regarding metadata and dedicated directory in the wider context: they all get in the way of this striving for simplicity, as otherwise it would be unreasonable to have all this code in the interface class. While we can argue about the need for metadata and dedicated directory, this whole approach chains our legs. If for instance we will be required to add support for more flags (something our design supports now), we will either have to add code in the interface class, or look for the functionality in the underlying file systems. Write once is a good example. While I'm not sure it can be done in the existing ones, no one promises us it'll be supported in future file systems.
But that's not all of it. Adding this functionality to the base class, as suggested here, would have a much larger impact on all current file systems. We may add bugs there (no one is bug proof) and will certainly need to widen our current tests to find these potential bugs. I'm not even sure how to do it - we currently have separate tests for FatFS and LittleFS, so we'll have to restructure our tests to cover it, something I don't think we want to do (especially in this time frame).

davidsaada replied:
As for the dedicated directory: the way I see it, one needs to decide what we want - to dedicate the file system for key value storage or give the option not to. If you want it dedicated for KV storage, then what are the other POSIX APIs needed for? One can even use TDBStore instead. Number of keys is no obstacle (as you suggested below). You also mentioned above that the user may want to have separate storage blocks for different uses (file system vs. KVStore). That's true if the user has enough space for it. However, I can think of use cases they don't have this space (for instance if they only use internal flash), meaning that separating the file systems is not optimal. Current design offers the flexibility of either having separate file system for POSIX and KVStore usage or for using both in the same file system, which is the second option. Now, if we want our file system to support both, we certainly have to separate these on the file system, hence the KVStore directory. Otherwise, iterator actions would return files that are not of the KVStore type (like logs), while the reset action wouldn't know which files to delete.

geky replied:
Cool, it sounds like we're on the same page now : )

> FileSystem is just an interface class, meaning that ALL APIs will need implementation in this header file.

Interface classes can have cpp files, FileSystem actually already has this along with several non-trivial defaults:
https://github.com/ARMmbed/mbed-os/blob/master/features/storage/filesystem/FileSystem.cpp#L117

> Adding this functionality to the base class, as suggested here, would have a much larger impact on all current file systems. We may add bugs there (no one is bug proof) and will certainly need to widen our current tests to find these potential bugs.

This is the same as adding FileSystemStore isn't it? I'm thinking FileSystem->KVStore inheritance is actually the easier solution.

> we'll have to restructure our tests to cover it

We're already planning on adding KVStore tests aren't we? Then we can just add FAT and KVStore to the list of KVStores being tested. That's the nice thing about inheritance, we don't create new tests.

Though we may end up catching more bugs, but that's no different than if we added FileSystemStore with tests.

davidsaada replied:
> Interface classes can have cpp files, FileSystem actually already has this along with several non-trivial defaults:

Sorry, missed this one.

> This is the same as adding FileSystemStore isn't it?

It's not the same. If we have FileSystem inherit KVStore, then all file systems will get the KVStore APIs by default, which is a change of behavior in current file systems, requiring tests and potentially adding bugs. Not the case with FileSystemStore.

geky replied:
> As for the dedicated directory: the way I see it, one needs to decide what we want

Ah, and I don't think that should be us. I think our users should be the ones who decide.

My proposal is we provide both:
1. FileSystemStore - provides dedicated directory
2. FileSystem->KVStore inheritance - provides direct access to filesystem as a KVStore

To be clear, I don't think option 1 is required for 5.11 and could be pushed back to 5.12. I also don't think providing only option 1 is a good idea. I'm worried about the cost of the FileSystemStore layer, even if we drop the extra metadata, I don't see how we can avoid dynamic allocation to set the directory's prefix. Also all of the logic and tests for option 2 are required for option 1, so I don't think we gain much by not providing option 2.

> to dedicate the file system for key value storage

Note with option 2 you can still use POSIX + KV APIs. Going back to the firmware + KV storage, the user could store all KV pairs as "kv.something", and the firmware as "fw.firmware1". This would let them store different types of files on the same disk.

This also avoids the dynamic string costs as the names are simple constants stored in the upper layers.

The same problem happens if you have a SecureStore on TDBStore, now the user may have files with different levels of enc/auth/rbp.

> Otherwise, iterator actions would return files that are not of the KVStore type (like logs)

Isn't this the purpose of the pattern in the iterator functions?

> If you want it dedicated for KV storage, then what are the other POSIX APIs needed for?

There are reasons besides the API that users may want to use something other than TDBStore.

- Maybe user wants KV store on NAND and needs to avoid bad blocks on storage?
- Maybe user wants to use FRAM or SD cards and doesn't want the overhead of simulating flash?
- Maybe user wants to plug the SD card into his machine to debug?
- Maybe user has so many KV pairs the RAM table gets unwieldy? 
- Maybe user is already using a filesystem elsewhere in the system and doesn't want to pay double the code cost to store keys on a separate piece of storage?

I'm not saying won't use TDBStore, there are plenty of use cases where TDBStore is a better choice. But we need to give users that choice. IoT is a developing field, if we can't provide flexibility we'll just lose users to other OSs.

geky replied:
Also consider if we end up adding enough features to TDBStore to where it can actually provide the POSIX APIs. I think the only way to introduce POSIX APIs in TDBStore in a backwards compatible would be by adding to the inheritance list.

geky replied:
Oh sorry! GitHub doesn't update in realtime so I missed your response.

> It's not the same. If we have FileSystem inherit KVStore, then all file systems will get the KVStore APIs by default, which is a change of behavior in current file systems, requiring tests and potentially adding bugs. Not the case with FileSystemStore.

Adding KVStore APIs shouldn't change the FileSystem APIs, does it? The only testing change is the addition of the KVStore APIs for LittleFS and FAT, which is true if we add the FileSystemStore or FileSystem->KVStore inheritance.

geky replied:
I think this is the sort of thing where code explains more than words. See https://github.com/davidsaada/mbed-os/pull/10.

We could have a really nicely bundled storage stack.

geky commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,378 @@
+
+### Data layout
+
+When storing the data, it is stored with a preceding 16-byte metadata header. Metadata includes flags and other parameters for basic validity checks.  
```
There was a conversation offline about this metadata. To be clear, this is unnecessary and adds overhead. All of the supported filesystems actually support the WRITE_ONCE flag so we just need to pass this flag to the file operations.

davidsaada replied:
Right. We first need to decide whether we want the write once feature or not for this class. It is now included for compatibility with `TDBStore` and `SecureStore`.

dannybenor replied:
@geky 
We need to understand "All of the supported filesystems actually support the WRITE_ONCE flag" because we do not know how to do that.
This flag is useful for having uniform tests over all implementations of kv_store interface. I do not see that somebody will use files in external devices that were encrypted in the device, therefore I do not see that adding this metadata has any bad side effect.

geky replied:
That's true in the case of SecureStore, and I don't see a downside to adding metadata in SecureStore. But SecureStore is an optional layer. Outside of SecureStore, it's completely reasonable for a user to mount a file system as a KVStore for the simpler API or to achieve some sort of compatibility.

davidsaada replied:
I'm not sure I understand how write once can be supported in current file systems. Furthermore, I don't think we can count on future file systems to support it. Even if we decide not to support it in FileSystemStore, having no metadata means we can't support ANY flags in the future, if required to.

geky replied:
One nice thing about flags is that we can always add features (until we run out of bits). If a user opens/sets a file with a flag the filesystem doesn't support, it should return an error.

We already have flags with partial support, for example the AUTHENTICATE_FLAG. If a user didn't read the documentation they may pass AUTHENTICATE_FLAG to TDBStore, in which case it should probably return an error.

If a feature isn't available in a specific filesystem, that's a limitation of the filesystem and the user should chose a different filesystem with the feature they're looking for.

Tangentially: We should add something like this to the handbook. Wikipedia has this for their big-ol-table of filesystems. Would probably help users:

| Storage Solution | KVStore API | FileSystem API | Power-loss resilience | Dynamic Wear-leveling | Static Wear-leveling | PC Support | Write-once | Encryption | Authentication | Rollback protection |
|------------------|-------------|----------------|-----------------------|-----------------------|----------------------|------------|------------|------------|----------------|---------------------|
| TDBStore | ✔️ | ❌ | ✔️ | ✔️ | ✔️ | ❌ | ✔️ | ❌ | ❌ | ❌ |
| LittleFS | ✔️ | ✔️ | ✔️ | ✔️ | ❌ | ❌ | ✔️ | ❌ | ❌ | ❌ |
| FAT | ✔️ | ✔️ | ❌ | ❌ | ❌ | ✔️ | ✔️ | ❌ | ❌ | ❌ |
| SecureStore | ✔️ | ❌ |  |  |  | ❌ | ✔️ | ✔️ | ✔️ | ✔️ |

---

We don't have to add WRITE_ONLY right now, we could error on storage that isn't TDBStore or SecureStore, and add WRITE_ONLY in 5.12.

Though WRITE_ONLY support shouldn't be that difficult given that both FAT and LittleFS have on disk structures that support it (though it's not exposed in the FileSystem API yet).

FAT:
``` cpp
FATFileSystem::file_open(fs_file_t *file, const char *path, int flags) {
    // blablabla
    f_open(fh, fpath, openmode);
    if (flags & WRITE_ONCE) {
        f_chmod(fpath, AM_RDO, AM_RDO);
    }
    // blablabla
}
```

LittleFS:
``` cpp
LittleFileSystem::file_open(fs_file_t *file, const char *path, int flags) {
    // blablabla
    f->cfg = NULL;
    if (flags & WRITE_ONCE) {
        // this could be more efficient, though may be less readable
        uint32_t *mask = new uint32_t;
        *mask = 0444;
        f->cfg = new lfs_file_config;
        f->cfg->attrs = new lfs_attr;
        f->cfg->attrs->type = 0xff;
        f->cfg->attrs->buffer = mask;
        f->cfg->attrs->size = sizeof(uint32_t);
        f->cfg->attrs->next = NULL;
    }
    lfs_file_opencfg(&_lfs, f, path, f->cfg);
    // blablabla
}
```

davidsaada replied:
> If a user didn't read the documentation they may pass AUTHENTICATE_FLAG to TDBStore, in which case it should probably return an error.

I'm pretty sure I wrote somewhere in the doc (will look for it), but any unknown flag is ignored. This is important for KVStore stacking (all flags are transferred to the lower level as is).
Regarding the table - I like it, but this one is misleading. SecureStore over FAT is not power-loss resilient while over TDBStore is.

geky replied:
> I'm pretty sure I wrote somewhere in the doc (will look for it), but any unknown flag is ignored.

I don't think this is a good idea. A user might right the following code and never realize their storage isn't secure:
``` cpp
SPIFBlockDevice spif(blabalbla);
TDBStore store(&spif);

store.set("secret-key", blababla, ENCRYPT_FLAG);
store.get("secret-key", blablabla);
```

We should be able to mask out flags as we pass the flags to lower layers. Is there other reasons to pass these flags down?

> SecureStore over FAT is not power-loss resilient while over TDBStore is.

Ah woops! I left those blank and then filled in all blanks with xs without thinking. Should be updated : )

geky commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,378 @@
+    Mutex _mutex;
+    size_t *_max_keys;
+    size_t *_num_keys;
+    bool _is_initialized;
```
~Building on previous comment, I expect FileSystemStore to have no state. The underlying FileSystem should by able to provide the storage behavior needed by the KVStore API or we've designed something wrong.~

Oh, this is implemented assuming there is a prefix for the directory the store is located in isn't it? Hmm...

davidsaada replied:
Yes it is. It's a hard coded directory name.

davidsaada replied:
Please also see comment above.

geky commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,378 @@
+- if `_is_initialized` return OK
+- Create and take `_mutex`
+- Create the FileSystemStore directory if it doesn't exist
+- Iterate on all files in this directory using `opendir` and `readdir` APIs, and store number of files in `_num_keys` 
```
I think I'm missing something, what's the point of `_num_keys`?

davidsaada replied:
Its purpose is to limit the number of stored keys. Just like the write once flag, it's an artificial requirement aiming for compatibility with `TDBStore`. It is actually written in the open issues section in the end of this document.

davidsaada replied:
OK, after discussing it with @dannybenor we decided to remove it. No point having this limitation here as well. Will change the doc.

geky commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,378 @@
+
+Pseudo code:  
+- Take `_mutex`
+- Delete all files under the FileSystemStore directory 
```
Shouldn't reset be a 1 to 1 mapping to FileSystem::reformat?

davidsaada replied:
Please see comment above.

geky commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,378 @@
+Pseudo code:  
+- if not `_is_initialized` return "not initialized" error
+- Take `_mutex` 
+- Using the `stat` API, extract file size
```
This may work itself out if we're changing this implementation, but you shouldn't need to call `stat` for the file size. This would trigger multiple file lookups when you can just open the file and access the size from there.

davidsaada replied:
Modified to use seek to end and then tell.

geky replied:
Or you could just use [File::size](https://github.com/ARMmbed/mbed-os/blob/master/features/storage/filesystem/File.h#L127)

geky commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,378 @@
+- Find file `key` under the FileSystemStore directory. If not existing, increase `_num_keys` by 1.
+- Take `_mutex`  
+- Allocate an `inc_set_handle_t` structure into `handle`
+- Duplicate `key` in `handle`.
```
Why do we need key in handle? Can't this just inc_set_handle_t just be a file?

davidsaada replied:
Not sure I understand. Do you suggest to store all handle fields in a dedicated file?

geky replied:
Oh, I misunderstood and thought this line involved a strcpy. This looks good :+1:

geky commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,378 @@
+		- Copy file name to `key` and return OK
+	- Using the FileSystem `readdir` API on handle's `dir_handle` field, read next file in directory
+- Return "not found" error
+- Release `_mutex`
```
Iterator is looking good 👍 Though not sure _mutex is needed

davidsaada replied:
Mutex locking can indeed be removed with the risk of getting erroneous results if keys are added.

davidsaada replied:
Ah sorry, what I wrote was a mistake. Mutex is indeed not needed here (as it won't protect us against those erroneous cases).

geky commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
````diff
@@ -0,0 +1,378 @@
+```C++
+
+// Instantiate fsstore with 64 as max number of keys
+FileSystemStore fsstore(64);
````
Doesn't FileSystem get passed to FileSystemStore at some point?

davidsaada replied:
Please see comment below. It will be passed now - will change the doc.

geky commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,378 @@
+### Open issues
+
+- Need to figure a way to prevent mutex abuse in incremental set APIs. 
+- Should we limit number of keys (as current design does) or not?
```
Do we limit keys in any other KVStore?

davidsaada replied:
Yes, in TDBStore (see comment above).

geky commented on docs/design-documents/features/storage/FileSystemStore/FileSystemStore_design.md:
```diff
@@ -0,0 +1,378 @@
+- Using the `stat` API, extract file size
+- Call `fopen` API on `key`, read mode to achieve a file handle
+- If failed, release `_mutex` and return "not found" error
+- Using `fread` API, read into a `key_metadata_t` structure
```
We shouldn't be calling retarget functions from FileSystemStore. The retarget layer is built entirely on top of the KVStore+FileSystem layer. The retarget layer pulls in a lot of toolchain logic and should generally be avoided until the last minute.

It's a feature for users but not for us, we also give the users the option to bypass the retarget layer to avoid the code cost. But that only works if we don't call the retarget layer internally.

davidsaada replied:
I agree. Actually the original design used the file system class APIs (and the file system was passed in the constructor). @dannybenor @offirko What do you think?

davidsaada replied:
OK, discussed it internally and we agree. Will revert to the original design.

geky reviewed (commented):
More comments. Thinking about the current requirements for FileSystemStore, there seems to be a conflict:
1. Thin wrapper for FileSystem -> KVStore (no state, minimal logic)
1. Provides KVStore in subdirectory

These two don't work together, moving KVStore to a subdirectory requires code and dynamic memory to hold the directory location.

It looks like we need to split these two requirements into separate implementations. We also need to enforce the statelessness of the adapter from FileSystem -> KVStore.

Given that FileSystem/KVStore exist at the same layer and this wrapper is very small, we should add KVStore to the inheritance chain and provide the adapter as default functions in FileSystem. This connects the dots between FileSystem and KVStore while also allowing FileSystemStore to be created on a subdirectory.

![image](https://user-images.githubusercontent.com/1075160/47060676-34713700-d194-11e8-84e2-846e7cf2e82d.png)

geky commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,460 @@
+
+## Design basics
+
+KVStore defines a simple, key value store like API set via this interface class. Classes implementing this interface store pairs of keys and values, where the keys are represented as character strings and the values are represented as binary blobs. Core APIs here are *get* and *set*, providing read and write access by key to the value in a single call. *remove* completes the set of core APIs. Thi ...
```
> KVStore defines a simple, key value store like API set via this interface class. Classes implementing this interface store pairs of keys and values, where the keys are represented as character strings and the values are represented as binary blobs. Core APIs here are *get* and *set*, providing read and write access by key to the value in a single call. *remove* completes the set of core APIs.

This is a great description 👍 

> This simple interface allows the implementing classes to have better performance, flash wear leveling and smaller footprint (of both code and data).  

This isn't really true as a consequence of the API. All the limited API gives you is a smaller code footprint. And the API you described in the previous sentence (no incremental write) is costly in terms of data.

davidsaada replied:
OK, will rephrase.

geky commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,460 @@
+### Derived implementations
+![KVStore Classes](./KVStore_classes.jpg)
+
+KVStore currently has a few derived implementations. [TDBStore](../TDBStore/TDBStore_design.md) should be chosen as the default solution, as it gives the best performance, flash wear leveling and lowest overhead. If we already have a file system and don't wish to have an additional storage solution, or if specific POSIX features (like file seeking) are required, then [FileSystemStore](../FileSyst ...
```
suggestion: This may work better as a bulleted list.

> [TDBStore](../TDBStore/TDBStore_design.md) should be chosen as the default solution, as it gives the best performance, flash wear leveling and lowest overhead.

This isn't true, unless you add the note "for small sets of data".

The decision on what type of storage solution to use really comes down to the type/size of data users are storing. This should probably be explained in the documentation going into the handbook later.

davidsaada replied:
"Small sets" can be misleading. 64 files seems like a good number for "TDBStore", but maybe also for LittleFS? Anyway, will remove this comment, as it is indeed more appropriate in Handbook.

geky replied:
Agreed, we'll need to write a good doc page for this.

I haven't tested it, but 64 files of ~4KiB is about where LittleFS would become a better choice over TDBStore. Those files are probably unreasonable for a KV store, but storage is large and users do try to store values that big (someone even hit LittleFS's 2GiB file limit already).

geky commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,460 @@
+ 
+// Core API
+int kv_set(const char *full_name_key, const void *buffer, size_t size, uint32_t create_flags);
+int kv_get(const char *full_name_key, void *buffer, size_t buffer_size, size_t *actual_size);
```
It's a bit tricky, but you can combine defaults + C compatibility in the same files. You can see how it's done over here:
https://github.com/ARMmbed/mbed-os/blob/master/features/netsocket/nsapi_dns.h#L103-L104

davidsaada replied:
If you refer to the `offset` argument in `get` API, then it is deliberately left out. This argument only serves the upper layers as a means of extracting chunks of data, parallel to the incremental set APIs. The latter were intentionally left out and so is this argument, as they don't serve the user.

geky replied:
I was also thinking of the optional "actual_size" parameter if we go that route. Don't feel like you need to though, it's up to you.

geky commented on docs/design-documents/features/storage/KVStore/KVStore_design.md:
```diff
@@ -0,0 +1,460 @@
+typedef struct {
+    KVStore *kvstore_intance;
+    KVStore::set_handle_t *set_handle;
+} inc_set_handle_t;
```
Should these structs have a `kv_` prefix?

davidsaada replied:
Good idea. Will change.

geky commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,448 @@
+
+### Requirements and assumptions
+
+SecureStore assumes that the underlying KVStore instances are instantiated and initialized. 
```
I would expect this to match the behaviour of FileSystem -> BlockDevice. Actually this is related to above comment.

davidsaada replied:
Please see my comment above.

geky replied:
I'm still expecting SecureStore to match the behaviour of SlicingBlockDevice -> BlockDevice.

Consider:
``` cpp
TDBStore kv;
SecureStore secure_kv(&kv);
secure_kv.init();
```

Do we need multiple init calls in this case? Does this mean the target needs to call all inits in the storage stack before providing KVStore::get_default_instance?

davidsaada replied:
Didn't know we have `get_default_instance` API for KVStore - maybe we should add it. Regardless, instantiating the KVStore class is something that should be implemented by the application init code (what we call the "integration code"). This code should instantiate the block device(s) and file system(s), so having an additional init call doesn't seem excessive. However, it's not `init` that bothers me like `deinit`. While calling file system/block device `init` by KVStore may be harmless calling `deinit` automatically may be more dangerous if the file system/block device is used for other purposes (which also relates to our discussion above). This was partially solved for block devices (with the init ref count), but will also need to get resolved for file systems if we follow this approach.

geky replied:
Irrespective of if we chain inits or not, we should have consistent API patterns in BlockDevice, KVStore, and (eventually) FileSystem.

Currently standardizing on chaining inits seems like the easiest option.

I'm not sure how we could move away from chained inits in BlockDevice without breaking user's code.

---

> Didn't know we have get_default_instance API for KVStore - maybe we should add it. 

Well we have FileSystem::get_default_instance and BlockDevice::get_default_instance so it sounds like a good idea : )

> This code should instantiate the block device(s) and file system(s), so having an additional init call doesn't seem excessive. 

get_default_instance is similar to class constructors, it doesn't return errors. So we should be avoiding init calls and leaving that up to the user.

The current behavior of BlockDevice::get_default_instance does this:
``` cpp
BlockDevice *bd = BlockDevice::get_default_instance();
bd->init(); // must call init
```

The behavior for FileSystem::get_default_instance is weird, just like the FileSystem constructor. I think we should improve this in the future.

---

Is there any issues with using the ref count in KVStore and FileSystem? This provides the most resilience to incorrect init calls from users.

davidsaada replied:
I tend to agree with everything you write here. There's no real issue with implementing init ref count   on file systems, other than time constrains. This is probably no biggie. Real problem is mainly with `deinit`, which isn't really used in real life, but more on test, so we may postpone it to 5.12.

geky replied:
Sounds like a plan 👍

geky replied:
Just to clarify, is the current plan non-reference counting for 5.11, and reference counting in 5.12 (or later) ?

geky commented:
Sorry for not being able to get through it all yet. Will have more review tomorrow.

geky commented on docs/design-documents/features/storage/TDBStore/TDBStore_design.md:
```diff
@@ -0,0 +1,497 @@
+Pseudo code:  
+- if `_is_initialized` return OK
+- Take `_mutex` 
+- Allocate `_ram_table` as an array of `_num_keys`  
```
It may be worth noting up in the RAM table section that the RAM table is fixed size at initialization time.

That being said, couldn't you make this RAM table dynamic with amortized doubling? This can be done pretty cheaply:
``` cpp
// constructor, optional initial table size, defaults to 16 keys
TDBStore::TDBStore(int initial_table_capacity = 16) {
    _ram_table = new key_struct[initial_table_capacity];
    _ram_table_capacity = initial_table_capacity;
    _ram_table_size = 0;
}

// ram-table insert function
void TDBStore::ram_table_insert(int pos, key_struct key) {
    if (_ram_table_size == _ram_table_capacity) {
        // double the ram_table
        int new_table_capacity = 2*_ram_table_capacity;
        key_struct *new_table = new key_struct[new_table_capacity];
        // copy over table contents
        memcpy(new_table, _ram_table, _ram_table_capacity * sizeof(key_struct));
        memset(new_table + _ram_table_capacity, 0, _ram_table_capacity);
        // free and replace old table
        delete[] _ram_table;
        _ram_table = new_table;
        _ram_table_capacity = new_table_capacity;
    }

    // insert new key
    memmove(&_ram_table[pos+1], &_ram_table[pos],
        (_ram_table_size-pos) * sizeof(key_struct));
    _ram_table[pos] = key;
    _ram_table_size += 1;
}

// destructor deletes keys, no other logic needed for removes (most tables don't decrease in size)
TDBStore::~TDBStore() {
    delete[] _ram_table;
}
```

davidsaada replied:
Yes it is an option. Only down size is the peak memory allocation. If we have 64 entries (8 bytes each), then adding a new key would result in a peak allocation of 1KB for this table, comparing to half as much if no doubling is implemented. But otherwise it's the better choice of course.

geky replied:
Note: This is just a suggestion, not a blocker. Can be added in 5.12.

If you let the user pass the capacity in the constructor, then you can avoid over-allocating if the user knows how many keys they will have. So I think the only downside is code size.

geky commented on docs/design-documents/features/storage/TDBStore/TDBStore_design.md:
```diff
@@ -0,0 +1,497 @@
+char buf[8];
+
+KVSTore::set_handle_t handle;
+res = tdbstore.set_start(&handle, "Key2", data_size, 0);
```
When this lands on the handbook, we may want an additional example that just shows off "get/set/remove". That would probably be the best way to show the simplicity of KVStore API and would make the best 1st impressions.

You're probably alread thinking that, but just wanted to mention it. We should have both examples since incremental set is also useful.

geky commented on docs/design-documents/features/storage/TDBStore/TDBStore_design.md:
```diff
@@ -0,0 +1,497 @@
+**reserved_space_set function**
+
+Header:  
+`virtual int reserved_space_set(void *data);`
```
Out of curiousity what's the reserved space used for? Is it for system-level attributes?

davidsaada replied:
Mainly for ROT/Device key. These will be used by the boot loader, and we don't want it to have all TDBStore code tucked in.

geky replied:
Ah, that makes sense. Is the reserved space at the end of the block? Does that mean we should be storing the ROT at the end of whichever block it resides in?

Interesting note: This same trick is used for FAT superblocks + MBR tables.

davidsaada replied:
It's actually at the beginning. Makes everything easier (also for the bootloader)

geky replied:
Oh ok. I guess we should actually make the ROT configurable at the byte-level in case we want to hide it in the actual bootloader image.

geky commented on docs/design-documents/features/storage/TDBStore/TDBStore_design.md:
```diff
@@ -0,0 +1,497 @@
+
+### Open issues
+
+- Need to figure a way to prevent mutex abuse in incremental set APIs. 
```
I think this is just a risk of using the advanced increment set APIs. May be worth a big warning note in the doxygen on the functions.

Also we should require that all KVStore implementations MUST hold a mutex if they're thread safe. Even the FileSystemStore where it's not strictly necessary. Otherwise KVStore <-> FileSystemStore isn't portable.

May be worth noting in the KVStore design doc.

davidsaada replied:
Good idea.

geky commented on features/storage/kvstore/KVStore.h:
```diff
@@ -0,0 +1,199 @@
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size = NULL, size_t offset = 0);
```
Here's an interesting thing I just found out.

Default arguments are not easily optimized out by compilers. The way they work is unfortunately hard for compilers to work around : (

Consider the case where a get with an offset is expensive, but a get with default offset is cheap (in terms of code size). Default arguments aren't different functions, so can't be gced by the compiler. The other trick a compiler can do, constant propagation, can't kick in unless LTO is on because this is a non-static function. So even if a user never uses a non-zero offset, they will still have to pay for all the offset logic.

I don't know if offset is expensive or not, but it's something to keep in mind. If a default is expensive it's probably better as an additional overloaded function.

And unfortunately for interface classes, the interface has to make the decision between overloads vs defaults, the implementation class can't add virtual functions easily.

geky commented on features/storage/kvstore/KVStore.h:
```diff
@@ -0,0 +1,199 @@
+     *
+     * @param[in]  it                   Iterator handle.
+     * @param[in]  key                  Buffer for returned key.
+     * @param[in]  key_size             Key buffer size.
```
Question, does key_size need to include the null-terminator? So would a key of size 32 fit in a buffer of size 32 or 33?

davidsaada replied:
Should have room for the null terminator as well.

geky commented on features/storage/kvstore/KVStore.h:
```diff
@@ -0,0 +1,199 @@
+     *
+     * @returns 0 on success or a negative error code on failure
+     */
+    virtual int iterator_open(iterator_t *it, const char *prefix = NULL);
```
Are we planning on adding a KVIterator/KVSet class to allow access to these functions through C++ classes?

We could provide such a class this way (without changing any of the implementations):
``` cpp
class KVStore {
    virtual int iterator_open(iterator_t *it, const char *prefix = NULL);
    virtual int iterator_next(iterator_t it, char *key, size_t key_size);
    virtual int iterator_close(iterator_t it);

    class Iterator {
    private:
        KVStore *_kvstore;
        const char *_prefix;
        iterator_t _it;
    public:
        Iterator Iterator(KVStore *kvstore, const char *prefix = NULL)
            : _kvstore(kvstore), _prefix(prefix), _it(NULL) {
        }

        virtual int open() {
            return _kvstore->iterator_open(&_it, _prefix);
        }

        virtual int next(iterator_t it, char *key, size_t key_size) {
            return _kvstore->iterator_next(&_it, key, key_size);
        }

        virtual int close(iterator_t it) {
            return _kvstore->iterator_close(&_it);
        }
    };
}
```

Usage:
``` cpp
TDBStore kvstore;
KVStore::Iterator it(&kvstore);
char key[KVStore::MAX_KEY_SIZE+1];

kvstore.init();
it.open();
it.next(key, KVStore::MAX_KEY_SIZE);
it.close();
```

Thoughts?

Commit:
Add required namespace instead of relying on mbed.h

mbed.h has 'using namespace mbed;', hence some of the files skip
adding required namespace to the code, it is always good to specify the
namespace around the elements in header and `using namespace` in
specific CPP files, instead of all

dreemkiller commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,448 @@
+SecureStore is a storage class, derived from KVStore. It adds security features to the underlying key value store.
+ 
+As such, it offers all KVStore APIs, with additional security options (which can be selected using the creation flags at set). These include:  
+- Encryption: Data is encrypted using the AES-CTR encryption method, where the IV consists of a randomly generated 8-byte nonce and an 8-byte counter. Key is derived from [Device Key](../../../../../../mbed-os/features/device_key/README.md), where salt is the key trimmed to 32 bytes, with "ENC" as prefix.  
```
What algorithm is used for the key derivation?

davidsaada replied:
We simply use the key derivation API `get_derived_key` from the `DeviceKey` feature, which implements the NIST SP 800-108 KDF in counter mode spec.

dreemkiller replied:
cool

dreemkiller commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,448 @@
+ 
+As such, it offers all KVStore APIs, with additional security options (which can be selected using the creation flags at set). These include:  
+- Encryption: Data is encrypted using the AES-CTR encryption method, where the IV consists of a randomly generated 8-byte nonce and an 8-byte counter. Key is derived from [Device Key](../../../../../../mbed-os/features/device_key/README.md), where salt is the key trimmed to 32 bytes, with "ENC" as prefix.  
+- Authentication: A 16-byte CMAC is calculated on all stored data (including metadata) and stored at the end of the record. When reading the record, calculated CMAC is compared with the stored one. In case of encryption, CMAC is calculated on the encrypted data. The key used for generating the CMAC is derived from [Device Key](../../../../../../mbed-os/features/device_key/README.md), where salt i ...
```
again, what key derivation algorithm?

davidsaada replied:
Please see above.

dreemkiller commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,448 @@
+As such, it offers all KVStore APIs, with additional security options (which can be selected using the creation flags at set). These include:  
+- Encryption: Data is encrypted using the AES-CTR encryption method, where the IV consists of a randomly generated 8-byte nonce and an 8-byte counter. Key is derived from [Device Key](../../../../../../mbed-os/features/device_key/README.md), where salt is the key trimmed to 32 bytes, with "ENC" as prefix.  
+- Authentication: A 16-byte CMAC is calculated on all stored data (including metadata) and stored at the end of the record. When reading the record, calculated CMAC is compared with the stored one. In case of encryption, CMAC is calculated on the encrypted data. The key used for generating the CMAC is derived from [Device Key](../../../../../../mbed-os/features/device_key/README.md), where salt i ...
+- Rollback protection: (Requires authentication) CMAC is stored in a designated rollback protected storage (also of KVStore type) and compared to when reading the data under the same key.
```
What happens when the compare fails?

davidsaada replied:
We return an "RBP authentication failure". Will add to the doc.

dreemkiller commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,448 @@
+
+### Data layout
+
+When storing the data, it is stored with a preceding metadata header. Metadata includes flags and security related parameters such as nonce. The CMAC, calculated for authentication, is stored at the end of the data as it is calculated on the fly, so it can't be stored with the metadata.  
```
Is this a nonce or an IV?

davidsaada replied:
Nonce (please also see my explanation below).

dreemkiller commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,448 @@
+### Basic implementation concepts
+
+When setting the data, as the code can't construct a single buffer to store all data (including metadata and possibly encrypted data) in one shot, it will be done in chunks, using the incremental set APIs. As for get, it will use the offset argument to extract metadata, data & CMAC separately.  
+Rollback protection (RBP) keys are stored in the designated rollback protection storage, of KVStore type also. RBP keys are the same as the SecureStore keys. 
```
Maybe we need to make a distinction between cryptographic keys and KVStore keys. I had to read this section a couple of times

davidsaada replied:
Thanks. Will rephrase to make this clearer.

dreemkiller commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,448 @@
+    Mutex _mutex;
+    KVStore *_underlying_kv;
+    KVStore *_rbp_kv;
+    void *_entropy;
```
What's the `_entropy` member for?

davidsaada replied:
We use TLS APIs in order to generate the random nonce, For this purpose, we use `mbedtls_entropy_context` initialized once at init, and then used for randomly generating the nonce using `mbedtls_entropy_func`. `_entropy` is declared as `void *` in the header and then casted to `mbedtls_entropy_context` in the implementation in order not to expose the TLS types in the header file.

dreemkiller commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,448 @@
+**set_add_data function**
+
+Header:  
+`virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);`
```
Should add documentation on the `set_add_data()` function letting the user know that the data is not protected until the `set_finalize()` function is called

davidsaada replied:
It is important to state that the stored record is not valid until `set_finalized` is called. `set_add_data` only adds a partial data chunk to the data. One cannot attempt to read the data before `set_finalize` is called, as the call is protected by a mutex. If we get a power failure between these calls, reading the data would fail on authentication, as CMAC is written last. If encryption is required, then `set_add_data` will write the data encrypted, so there's no fear here that the written data is exposed.

dreemkiller replied:
correct. can we add that to the documentation of the functions?

davidsaada replied:
Sure. Will add.

dreemkiller commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,448 @@
+	- Decrement `data_size` by `chunk_size`
+- Call `_underlying_kv` `get` API with on a local `read_cmac` variable, size 16
+- Generate CMAC on local `cmac` variable 
+- If `read_cmac` doesn't match `cmac`, return "data corrupt error"
```
This compare needs to be a constant-time compare. This is harder than it sounds

davidsaada replied:
Please elaborate. Not sure I understand.

dreemkiller replied:
a naive compare would look something like this **NOTE: THIS IS BAD. DO NOT USE**:
```C
int pass = 1;
for (int i = 0; i < mac_len; i++) {
    if(mac[i] != reference_mac[i]) {
        pass = 0;
        break;
    }
}
return pass;
```

However, attacker could use this to measure time differentials in different attempts, and then glean information about the underlying key.
A better (but still not sufficient) implementation **NOTE: THIS IS STILL BAD DO NOT USE**:
```C
int pass = 1;
for (int i = 0; i < mac_len; i++) {
    if(mac[i] != reference_mac[i]) {
        pass = 0;
    }
}
return pass;
```
Notice that now it **should** take the same amount of time regardless of where the first miscompare is. However, some compilers may optimize this and create a problem.
MbedTLS has a constant time comparison function `mbedtls_ssl_safer_memcmp()` which should be used for this.

davidsaada replied:
Thanks for the clarification. Will use this function.

dreemkiller commented on docs/design-documents/features/storage/SecureStore/SecureStore_design.md:
```diff
@@ -0,0 +1,448 @@
+- Call `set_finalize`
+- Return OK
+
+**get function**
```
Should add documentation to the get function telling users that partial reads of encrypted/MACed data are inefficient, and that if they are going to eventually read all of the data, it's better to do the read in one shot, otherwise they will experience performance problems

davidsaada replied:
Thanks, will add that.

dreemkiller reviewed (commented):
There's a general confusion in this document and in the code between nonces and IVs. This design uses IVs, which is fine, but the terms are not interchangeable (IVs must be unpredictable, which generating randomly as is done here, is acceptable. Nonces do not need to be unpredictable)
The rest of the comments are just asking for more information in the document.

mudassar-ublox committed:
Added unsupported features for UBLOX targets

davidsaada commented:
@dreemkiller Thanks for taking the time to review this.
> There's a general confusion in this document and in the code between nonces and IVs. This design uses IVs, which is fine, but the terms are not interchangeable (IVs must be unpredictable, which generating randomly as is done here, is acceptable. Nonces do not need to be unpredictable)
The rest of the comments are just asking for more information in the document.

We use the AES-CTR method, where the IV consist of a random nonce and an incrementing counter (based on position in data). Maybe I used the wrong terms here, so will be glad if you correct me.

davidsaada commented:
> What does set do when bits in create_flags are not supported?

Bottom line - it ignores them. Here's the explanation (was also part of a long discussion we had here with @geky and @flit).
KVStore uses a cascading class model, meaning that a high layer class should call a lower layer one. `set` is no exception. SecureStore, for instance, can be implemented on top of TDBStore. Now, different classes can support different flags, and it's unreasonable to assume that each class knows which flags are supported by its underlying ones. So according to this model, the top layer class extracts the flags it knows, and sends the *full* flag mask to the underlying class (it can mask off its own supported flags, but this won't really change the behavior). That's why we can't return an error for unsupported flags.

dreemkiller commented:
I think we should remove the term `nonce` from the code and documentation of this feature.
What you are using here, **appropriately**, are Initialization Vectors (IVs). They are subtly different from nonces. The key difference being: nonces do not have to be unpredictable, IVs do. 
As far as I can tell, you are using things appropriately. Just change the names.

davidsaada commented:
> I think we should remove the term nonce from the code and documentation of this feature. What you are using here, appropriately, are  Initialization Vectors (IVs). They are subtly different from nonces. The  key difference being: nonces do not have to be unpredictable, IVs do. As far as I can tell, you are using things appropriately. Just change the names.

Thanks for the clarification. Would just like to make things clearer - what I keep in the metadata is not the entire IV, but just the first half (the random one). while the second half (the counter) is not kept. Can I still use the term "IV" to describe this (the random part)?

dreemkiller commented:
Yes, the first half, the random part, is an IV, not a nonce.

Commit:
Add QSPIF block device to default system storage

Commit:
fixes the credentials storage, copies them internally rather than just a pointer to external storage

Commit:
Merge pull request #8371 from u-blox/ublox-cellular-unsupported-features

Cellular: Add unsupported features for UBLOX targets

Commit:
Merge pull request #8550 from deepikabhavnani/mbed_h_fixes

Add required namespace instead of relying on mbed.h

Commit:
Merge pull request #8566 from u-blox/github8520issue

Wifi: ublox fix to keep the credentials stored

Commit:
Merge pull request #8563 from marcuschangarm/default_qspif

Add QSPIF block device to default system storage
